core/kernel/transfer_list.c

```diff
@@ -0,0 +1,580 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, Linaro Limited
+ */
+
+/*******************************************************************************
+ * Transfer list library compliant with the Firmware Handoff specification at:
+ * https://github.com/FirmwareHandoff/firmware_handoff
+ ******************************************************************************/
+
+#include <kernel/cache_helpers.h>
+#include <kernel/panic.h>
+#include <kernel/transfer_list.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <string.h>
+#include <util.h>
+
+/*******************************************************************************
+ * Adapt a physical address to match the maximum transfer entry data alignment
+ * required by an existing transfer list.
+ * Compliant with 2.4.6 of Firmware Handoff specification (v0.9).
+ * @pa: Physical address for adapting.
+ * @tl: Pointer to the existing transfer list.
+ * Return the adapted physical address.
+ ******************************************************************************/
+static paddr_t get_align_base_addr(paddr_t pa,
+				   struct transfer_list_header *tl)
+{
+	paddr_t align_mask = TL_ALIGNMENT_FROM_ORDER(tl->alignment) - 1;
+	paddr_t align_off = (paddr_t)tl & align_mask;
+	paddr_t new_addr = (pa & ~align_mask) + align_off;
+
+	if (new_addr < pa)
+		new_addr += TL_ALIGNMENT_FROM_ORDER(tl->alignment);
+
+	return new_addr;
+}
+
+static void unmap_list(struct transfer_list_header *tl, size_t sz)
+{
+	if (core_mmu_remove_mapping(MEM_AREA_TRANSFER_LIST, tl, sz))
+		panic("Failed to remove transfer list mapping");
+}
+
+struct transfer_list_header *transfer_list_map(paddr_t pa)
+{
+	struct transfer_list_header *tl = NULL;
+	size_t sz = SMALL_PAGE_SIZE;
+	size_t old_sz = 0;
+
+	while (true) {
+		tl = core_mmu_add_mapping(MEM_AREA_TRANSFER_LIST, pa, sz);
+		if (!tl) {
+			EMSG("Failed to map TL with PA %#"PRIxPA", size %#zx",
+			     pa, sz);
+			return NULL;
+		}
+		old_sz = sz;
+
+		if (transfer_list_check_header(tl) == TL_OPS_NONE) {
+			unmap_list(tl, sz);
+			return NULL;
+		}
+
+		if (tl->max_size <= sz)
+			return tl;
+
+		sz = ROUNDUP(tl->max_size, SMALL_PAGE_SIZE);
+		unmap_list(tl, old_sz);
+	}
+}
+
+void transfer_list_unmap_sync(struct transfer_list_header *tl)
+{
+	size_t sz = tl->max_size;
+
+	transfer_list_update_checksum(tl);
+	dcache_cleaninv_range(tl, sz);
+	unmap_list(tl, sz);
+}
+
+void transfer_list_unmap_nosync(struct transfer_list_header *tl)
+{
+	unmap_list(tl, tl->max_size);
+}
+
+void transfer_list_dump(struct transfer_list_header *tl)
+{
+	struct transfer_list_entry *te = NULL;
+	int i = 0;
```

Comment:
`__maybe_unused`

Could you build TL support with `CFG_TEE_CORE_LOG_LEVEL=0 CFG_TRANSFER_LIST=y CFG_WERROR=y`? I think it would have triggered an error here.

Reply:
Fixed.